diff --git a/simplesamlphp_auth.module b/simplesamlphp_auth.module
index 5c5f200..2ceff34 100644
--- a/simplesamlphp_auth.module
+++ b/simplesamlphp_auth.module
@@ -232,9 +232,28 @@ function simplesamlphp_auth_init() {
         $ext_user = user_external_load($authname);
         $skip_user_finalization = FALSE;
 
+        // Add authmaps on the fly for manually created users.
+        if (empty($ext_user) && variable_get('simplesamlphp_auth_addauthmaps', TRUE)) {
+
+          $context = array(
+            'simplesamlphp_auth_as' => $_simplesamlphp_auth_as,
+            'simplesamlphp_auth_saml_attributes' => $_simplesamlphp_auth_saml_attributes,
+            'simplesamlphp_auth_saml_config' => $_simplesamlphp_auth_saml_config,
+            'simplesamlphp_auth_saml_version' => $_simplesamlphp_auth_saml_version,
+            'authname' => $authname,
+          );
+
+          drupal_alter('simplesamlphp_auth_addauthmaps', $existing_user, $context);
+          if (!empty($existing_user)) {
+
+            user_set_authmaps($existing_user, array('authname_simplesamlphp_auth' => $existing_user->name));
+            $ext_user = $existing_user;
+          }
+        }
+
         if (!$ext_user) {
           // First we check the admin settings for simpleSAMLphp and find out if we are allowed to register users.
-          if (variable_get('simplesamlphp_auth_registerusers', TRUE)) {
+          if (variable_get('simplesamlphp_auth_registerusers', TRUE) && !$existing_user) {
 
             // We are allowed to register new users.
             _simplesaml_auth_debug(t('Register [%authname]', array('%authname' => $authname)));
@@ -540,6 +559,13 @@ function simplesamlphp_auth_settings() {
     '#description' => t('Determines wether or not the module should automatically create/register new Drupal accounts for users that authenticate using SimpleSAMLphp. Unless you\'ve done some custom work to provision Drupal accounts with the necessary authmap entries you will want this checked.<br /><br />NOTE: If unchecked each user must already have been provisioned a Drupal account with an appropriate entry in the authmap table before logging in. Otherwise they will receive a notice and be denied access. Be aware that simply creating a Drupal account will not create the necessary entry in the authmap table.'),
   );
 
+  $form['simplesamlphp_auth_grp_reg']['simplesamlphp_auth_addauthmaps'] = array(
+    '#type' => 'checkbox',
+    '#title' => t('Enable authentication for manually created users'),
+    '#default_value' => variable_get('simplesamlphp_auth_addauthmaps', TRUE),
+    '#description' => t('Allows users accounts created manually (not created by auto-provisoning) to authenticate using SimpleSAMLphp by adding the necessary authmap entries.'),
+   );
+
   $form['simplesamlphp_auth_grp_auth'] = array(
     '#type' => 'fieldset',
     '#title' => t('Drupal Authentication'),
@@ -1021,3 +1047,13 @@ function simplesamlphp_auth_get_attributes($attribute = NULL) {
   return $result;
 }
 
+/**
+ * Implements hook_simplesamlphp_auth_addauthmaps_alter().
+ */
+function simplesamlphp_auth_simplesamlphp_auth_addauthmaps_alter(&$existing_user, $context) {
+
+  if (empty($existing_user)) {
+
+    $existing_user = user_load_by_name($context['authname']);
+  }
+}
